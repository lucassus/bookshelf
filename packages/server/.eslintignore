coverage
dist
node_modules
jest.config.js
src/**/*.generated.ts
