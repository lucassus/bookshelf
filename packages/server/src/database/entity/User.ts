import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  OneToOne,
  JoinColumn,
  Index,
  CreateDateColumn,
  UpdateDateColumn,
  OneToMany
} from "typeorm";

import { Avatar } from "./Avatar";
import { BookCopy } from "./BookCopy";

@Entity({ name: "users" })
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column()
  @Index({ unique: true })
  email: string;

  @Column({ default: "" })
  info: string;

  @OneToOne(() => Avatar, { eager: true })
  @JoinColumn({ name: "avatar_id" })
  avatar: Avatar;

  @Column({ name: "avatar_id" })
  avatarId: number;

  @OneToMany(() => BookCopy, (bookCopy) => bookCopy.owner)
  ownedBookCopies: Promise<BookCopy[]>;

  @OneToMany(() => BookCopy, (bookCopy) => bookCopy.borrower)
  borrowedBookCopies: Promise<BookCopy[]>;

  @CreateDateColumn({ name: "created_at" })
  createdAt: Date;

  @UpdateDateColumn({ name: "updated_at" })
  updatedAt: Date;
}
