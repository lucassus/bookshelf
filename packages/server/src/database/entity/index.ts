export { Author } from "./Author";
export { Avatar } from "./Avatar";
export { Book } from "./Book";
export { BookCopy } from "./BookCopy";
export { User } from "./User";
