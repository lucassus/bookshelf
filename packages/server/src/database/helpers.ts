import { SecureId } from "./SecureId";

type EntityTypes = "Author" | "Book" | "User" | "BookCopy";

export const secureId = new SecureId<EntityTypes>({
  separator: "-"
});
