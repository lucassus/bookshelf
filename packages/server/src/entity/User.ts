import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  OneToOne,
  JoinColumn,
  Index,
  BaseEntity
} from "typeorm";

import { Avatar } from "./Avatar";

// TODO: Add created_at, updated_at

@Entity()
export class User extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column()
  @Index({ unique: true })
  email: string;

  @OneToOne(() => User)
  @JoinColumn({ name: "avatar_id" })
  avatar: Avatar;
}
