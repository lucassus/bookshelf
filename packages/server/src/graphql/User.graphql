type User {
  id: ID!
  name: String!
  info: String
  email: String!
  avatar: Avatar!
  ownedBookCopies: [BookCopy!]!
  borrowedBookCopies: [BookCopy!]!
}

type Avatar {
  image: Image!
  color: String!
}
