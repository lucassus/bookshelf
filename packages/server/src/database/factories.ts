import { getManager } from "typeorm";

import { Author } from "./entity/Author";
import { Avatar } from "./entity/Avatar";
import { Book } from "./entity/Book";
import { BookCopy } from "./entity/BookCopy";
import { User } from "./entity/User";

export const createAvatar = (attributes: Partial<Avatar> = {}) => {
  const manager = getManager();

  return manager.save(
    manager.create(Avatar, {
      imagePath: "/images/avatars/w13.png",
      color: "yellow",
      ...attributes
    })
  );
};

export const createUser = async (
  attributes: Partial<User> = { name: "Alice" }
) => {
  const manager = getManager();

  const { ...userAttributes } = attributes;

  if (userAttributes.name && userAttributes.email === undefined) {
    userAttributes.email = `${userAttributes.name.toLocaleLowerCase()}@email.com`;
  }

  if (userAttributes.avatarId === undefined) {
    const avatar = await createAvatar();
    userAttributes.avatarId = avatar.id;
  }

  const user = manager.create(User, userAttributes);

  return manager.save(user);
};

export const createAuthor = (attributes: Partial<Author> = {}) => {
  const manager = getManager();

  return manager.save(
    manager.create(Author, {
      name: "Andrzej Sapkowski",
      photoPath: "/images/book-authors/andrzej-sapkowski.jpg",
      ...attributes
    })
  );
};

export const createBook = async (
  attributes: Partial<Book> & { authorAttributes?: Partial<Author> } = {}
) => {
  const manager = getManager();

  const { authorAttributes, ...bookAttributes } = attributes;

  if (bookAttributes.authorId === undefined) {
    const author = await createAuthor(authorAttributes);
    bookAttributes.authorId = author.id;
  }

  return manager.save(
    manager.create(Book, {
      title: "Baptism of fire",
      coverPath: "/images/book-covers/witcher3.jpg",
      ...bookAttributes
    })
  );
};

export const createBookCopy = async (
  attributes: Partial<BookCopy> & { bookAttributes?: Partial<Book> } & {
    ownerAttributes?: Partial<User>;
  } = {}
) => {
  const manager = getManager();

  const { bookAttributes, ownerAttributes, ...bookCopyAttributes } = attributes;

  if (bookCopyAttributes.bookId === undefined) {
    const book = await createBook(bookAttributes);
    bookCopyAttributes.bookId = book.id;
  }

  if (bookCopyAttributes.ownerId === undefined) {
    const user = await createUser(ownerAttributes);
    bookCopyAttributes.ownerId = user.id;
  }

  return manager.save(manager.create(BookCopy, bookCopyAttributes));
};
