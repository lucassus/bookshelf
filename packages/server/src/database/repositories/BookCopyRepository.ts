import { AbstractRepository, EntityRepository } from "typeorm";

import { BookCopy } from "../entity";

@EntityRepository(BookCopy)
export class BookCopyRepository extends AbstractRepository<BookCopy> {
  async borrow(id: string | number, borrowerId: number) {
    const bookCopy = await this.repository.findOneOrFail(id);

    // TODO: Implement better errors handling
    // TODO: Display a toast on the client side
    if (bookCopy.ownerId === borrowerId) {
      throw new Error("Cannot borrow own book.");
    }

    if (bookCopy.borrowerId) {
      throw new Error("Cannot borrow this book copy. It is already borrowed.");
    }

    bookCopy.borrowerId = borrowerId;
    return this.repository.save(bookCopy);
  }

  async return(id: string | number, borrowerId: number) {
    const bookCopy = await this.repository.findOneOrFail({
      where: {
        id,
        borrowerId
      }
    });

    bookCopy.borrowerId = null;
    return this.repository.save(bookCopy);
  }
}
