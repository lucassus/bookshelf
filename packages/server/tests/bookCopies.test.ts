import { ApolloServer, gql } from "apollo-server-express";
import { createTestClient } from "apollo-server-testing";
import { getCustomRepository } from "typeorm";

import { User } from "../src/database/entity/User";
import { createBookCopy, createUser } from "../src/database/factories";
import { secureId } from "../src/database/helpers";
import { BookCopyRepository } from "../src/database/repositories/BookCopyRepository";
import { createServer } from "../src/server";

let currentUser: User;
let server: ApolloServer;

beforeEach(async () => {
  // TODO: Implement a better way for authenticating a user
  currentUser = await createUser();

  server = createServer();
});

test("borrow book copy", async () => {
  // Given
  const { query } = createTestClient(server);

  const bookCopy = await createBookCopy({
    bookAttributes: {
      title: "Time of contempt"
    },
    ownerAttributes: {
      name: "Bob",
      email: "bob@email.com"
    }
  });

  // When
  const res = await query({
    query: gql`
      mutation BorrowBookCopy($id: ID!) {
        borrowBookCopy(id: $id) {
          id
          book {
            id
            title
          }
          borrower {
            id
            name
          }
        }
      }
    `,
    variables: { id: secureId.toExternal(bookCopy.id, "BookCopy") }
  });

  // Then
  expect(res.data).not.toBeUndefined();
  expect(res.data!.borrowBookCopy).toMatchSnapshot();
});

test("return book copy", async () => {
  // Given
  const { query } = createTestClient(server);

  const bookCopy = await createBookCopy({
    ownerAttributes: { name: "Bob", email: "bob@email.com" }
  });
  await getCustomRepository(BookCopyRepository).borrow(
    bookCopy.id,
    currentUser.id
  );

  // When
  const res = await query({
    query: gql`
      mutation ReturnBookCopy($id: ID!) {
        returnBookCopy(id: $id) {
          id
          book {
            id
            title
          }
          borrower {
            id
            name
          }
        }
      }
    `,
    variables: { id: secureId.toExternal(bookCopy.id, "BookCopy") }
  });

  // Then
  expect(res.data).not.toBeUndefined();
  expect(res.data!.returnBookCopy).toMatchSnapshot();
});
