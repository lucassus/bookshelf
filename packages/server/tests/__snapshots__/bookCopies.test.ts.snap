// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`borrow book copy 1`] = `
Object {
  "book": Object {
    "id": "MS1Cb29r",
    "title": "Time of contempt",
  },
  "borrower": Object {
    "id": "MS1Vc2Vy",
    "name": "Rosalind Effertz",
  },
  "id": "MS1Cb29rQ29weQ==",
  "owner": Object {
    "id": "Mi1Vc2Vy",
    "name": "Bob",
  },
}
`;

exports[`return book copy 1`] = `
Object {
  "book": Object {
    "id": "MS1Cb29r",
    "title": "Minus",
  },
  "borrower": null,
  "id": "MS1Cb29rQ29weQ==",
  "owner": Object {
    "email": "bob@email.com",
    "id": "Mi1Vc2Vy",
    "name": "Bob",
  },
}
`;
